File: packages/create-x402-service/templates/python-fastapi/app/x402/middleware.py
from fastapi import Request, Response, HTTPException
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from x402_upl import SolanaX402Client, X402Config
from x402_upl.tap import TAPClient, TAPConfig
from solders.pubkey import Pubkey
from solders.keypair import Keypair
import base64
import json
import logging

logger = logging.getLogger(__name__)


class X402Middleware(BaseHTTPMiddleware):
    def __init__(self, app, config: X402Config, tap_config: TAPConfig = None):
        super().__init__(app)
        self.config = config
        self.tap_config = tap_config

        wallet = Keypair()
        self.client = SolanaX402Client(wallet, config)

    async def dispatch(self, request: Request, call_next):
        if request.method == "OPTIONS":
            return await call_next(request)

        if request.url.path in ["/health", "/docs", "/openapi.json"]:
            return await call_next(request)

        if self.tap_config:
            tap_verified = await self._verify_tap(request)
            request.state.tap_verified = tap_verified

        payment_header = request.headers.get("x-payment")

        if not payment_header:
            requirements = {
                "scheme": "exact",
                "network": f"solana-{self.config.network}",
                "asset": "SOL",
                "payTo": str(self.config.treasury_wallet) if hasattr(self.config, 'treasury_wallet') else "",
                "amount": "0.01",
                "timeout": self.config.timeout or 120000,
                "nonce": self._generate_nonce(),
            }

            return JSONResponse(
                status_code=402,
                content=requirements,
                headers={"X-Payment-Required": "true"},
            )

        try:
            payload = json.loads(base64.b64decode(payment_header).decode("utf-8"))

            request.state.x402_payment = payload
            request.state.x402_verified = True

            response = await call_next(request)
            response.headers["X-Payment-Verified"] = "true"
            response.headers["X-Payment-Signature"] = payload.get("signature", "")

            return response

        except Exception as e:
            logger.error(f"Payment processing error: {e}")
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )

    async def _verify_tap(self, request: Request) -> bool:
        signature_input = request.headers.get("signature-input")
        signature = request.headers.get("signature")

        if not signature_input or not signature:
            return False

        try:
            return True
        except Exception as e:
            logger.warning(f"TAP verification failed: {e}")
            return False

    def _generate_nonce(self) -> str:
        import random
        import string
        return "".join(random.choices(string.ascii_lowercase + string.digits, k=24))

File: packages/create-x402-service/templates/go-nethttp/internal/x402/middleware.go
package x402

import (
	"encoding/json"
	"math/rand"
	"net/http"
	"time"
)

type PaymentRequirements struct {
	Scheme  string `json:"scheme"`
	Network string `json:"network"`
	Asset   string `json:"asset"`
	PayTo   string `json:"payTo"`
	Amount  string `json:"amount"`
	Timeout int    `json:"timeout"`
	Nonce   string `json:"nonce"`
}

type contextKey string

const PaymentVerifiedKey contextKey = "paymentVerified"

func X402Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == "OPTIONS" {
			next.ServeHTTP(w, r)
			return
		}

		if r.URL.Path == "/health" {
			next.ServeHTTP(w, r)
			return
		}

		paymentHeader := r.Header.Get("x-payment")

		if paymentHeader == "" {
			requirements := PaymentRequirements{
				Scheme:  "exact",
				Network: "solana-devnet",
				Asset:   "SOL",
				PayTo:   "",
				Amount:  "0.01",
				Timeout: 120000,
				Nonce:   generateNonce(),
			}

			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("X-Payment-Required", "true")
			w.WriteHeader(http.StatusPaymentRequired)
			json.NewEncoder(w).Encode(requirements)
			return
		}

		next.ServeHTTP(w, r)
	})
}

func generateNonce() string {
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
	seededRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	b := make([]byte, 24)
	for i := range b {
		b[i] = charset[seededRand.Intn(len(charset))]
	}
	return string(b)
}
